pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -38,18 +38,110 @@ var (
+	if chain.FilterChainMatch == nil {
+		chain.FilterChainMatch = &listener.FilterChainMatch{}
+	}
+	listenerAddress := l.Address
+	if sockAddr := listenerAddress.GetSocketAddress(); sockAddr != nil {
+		chain.FilterChainMatch.DestinationPort = &types.UInt32Value{Value: sockAddr.GetPortValue()}
+		if cidr := util.ConvertAddressToCidr(sockAddr.GetAddress()); cidr != nil {
+			if chain.FilterChainMatch.PrefixRanges != nil && len(chain.FilterChainMatch.PrefixRanges) != 1 {
+				log.Debugf("Intercepted inbound listener %s have neither 0 or 1 prefix ranges. Actual:  %d",
+					l.Name, len(chain.FilterChainMatch.PrefixRanges))
+			}
+			chain.FilterChainMatch.PrefixRanges = []*core.CidrRange{util.ConvertAddressToCidr(sockAddr.GetAddress())}
+		}
+		insertOriginalListenerName(chain, l.Name)
+	}
+	for _, filter := range l.ListenerFilters {
+		if needTLS = needTLS || filter.Name == xdsutil.TlsInspector; needTLS {
+			break
+		}
+	}
+	return chain, needTLS
+}
+
+// Accumulate the filter chains from per proxy service listeners
+func reduceInboundListenerToFilterChains(listeners []*xdsapi.Listener) ([]*listener.FilterChain, bool) {
+	needTLS := false
+	chains := make([]*listener.FilterChain, 0)
+	for _, l := range listeners {
+		// default bindToPort is true and these listener should be skipped
+		if v1Opt := l.GetDeprecatedV1(); v1Opt == nil || v1Opt.BindToPort == nil || v1Opt.BindToPort.Value {
+			// A listener on real port should not be intercepted by virtual inbound listener
+			continue
+		}
+		for _, c := range l.FilterChains {
+			newChain, needTLSLocal := amendFilterChainMatchFromInboundListener(c, l, needTLS)
+			chains = append(chains, newChain)
+			needTLS = needTLS || needTLSLocal
+		}
+	}
+	return chains, needTLS
+}
+
+func (builder *ListenerBuilder) aggregateVirtualInboundListener() *ListenerBuilder {
+	// Deprecated by envoyproxy. Replaced
+	// 1. filter chains in this listener
+	// 2. explicit original_dst listener filter
+	// UseOriginalDst: proto.BoolTrue,
+	builder.virtualInboundListener.UseOriginalDst = nil
+	builder.virtualInboundListener.ListenerFilters = append(builder.virtualInboundListener.ListenerFilters,
+		&listener.ListenerFilter{
+			Name: xdsutil.OriginalDestination,
+		},
+	)
+	filterChains, needTLS := reduceInboundListenerToFilterChains(builder.inboundListeners)
+
+	builder.virtualInboundListener.FilterChains =
+		append(builder.virtualInboundListener.FilterChains, filterChains...)
+
+	if needTLS {
+		builder.virtualInboundListener.ListenerFilters =
+			append(builder.virtualInboundListener.ListenerFilters, &listener.ListenerFilter{
+				Name: xdsutil.TlsInspector,
+			})
+	}
+	return builder
 }
 
 func NewListenerBuilder(node *model.Proxy) *ListenerBuilder {
 	builder := &ListenerBuilder{
 		node: node,
+		// The extra inbound listener has no side effect for iptables that doesn't redirect to 15006
+		useInboundFilterChain: true,
```

Comment:
you dont need this option at all. The entier inbound listener is new [with virtual listener]. So lets go all in and have either the old style one virtual+all phsical, or one virtualOutbound+physical listenres, and one physicalInbound+filterChains

Reply:
Sounds reasonable - if the sidecar has enabled the redirect to 15006 we can generate only new style inbound. 

It is probably safer too (security-wise), and anyways the iptables would not send inbound to the new port.